src/retry.rs

```diff
@@ -0,0 +1,438 @@
+//! with including best practices by default, such as a retry budget.
+//!
+//! # Defaults
+//!
+//! The default retry behavior of a `Client` is to only retry requests where an
+//! error or low-level protocol NACK is encountered that is known to be safe to
+//! retry. Note however that providing a specific retry policy will override
+//! the default, and you will need to explicitly include that behavior.
+//!
+//! All policies default to including a retry budget that permits 20% extra
+//! requests to be sent.
+//!
+//! # Scoped
+//!
+//! A client's retry policy is scoped. That means that the policy doesn't
+//! apply to all requests, but only those within a user-defined scope.
+//!
+//! Since all policies include a budget by default, it doesn't make sense to
+//! apply it on _all_ requests. Rather, the retry history applied by a budget
+//! should likely only be applied to the same host.
+//!
+//! # Classifiers
+//!
+//! A retry policy needs to be configured with a classifier that determines
+//! if a request should be retried. Knowledge of the destination server's
+//! behavior is required to make a safe classifier. **Requests should not be
+//! retried** if the server cannot safely handle the same request twice, or if
+//! it causes side effects.
+//!
+//! Some common properties to check include if the request method is
+//! idempotent, or if the response status code indicates a transient error.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use tower::retry::budget::{Budget as _, TpsBudget as Budget};
+
+/// Builder to configure retries
+///
+/// Construct with [`for_host()`].
+#[derive(Debug)]
+pub struct Builder {
+    //backoff: Backoff,
+    budget: Option<f32>,
+    classifier: classify::Classifier,
+    max_retries_per_request: u32,
+    scope: scope::Scoped,
+}
+
+/// The internal type that we convert the builder into, that implements
+/// tower::retry::Policy privately.
+#[derive(Clone, Debug)]
+pub(crate) struct Policy {
+    budget: Option<Arc<Budget>>,
+    classifier: classify::Classifier,
+    max_retries_per_request: u32,
+    retry_cnt: u32,
+    scope: scope::Scoped,
+}
+
+//#[derive(Debug)]
+//struct Backoff;
+
+/// Create a retry builder with a request scope.
+///
+/// To provide a scope that isn't a closure, use the more general
+/// [`Builder::scoped()`].
+pub fn for_host<S>(host: S) -> Builder
+where
+    S: PartialEq<str> + Send + Sync + 'static,
+{
+    scoped(move |req| &host == req.uri().host().unwrap_or(""))
+}
+
+fn scoped<F>(func: F) -> Builder
+where
+    F: Fn(&Req) -> bool + Send + Sync + 'static,
+{
+    Builder::scoped(scope::ScopeFn(func))
+}
+
+// ===== impl Builder =====
+
+impl Builder {
+    /// Create a scoped retry policy.
+    ///
+    /// For a more convenient constructor, see [`retry::for_host()`].
+    pub fn scoped(scope: impl scope::Scope) -> Self {
+        Self {
+            budget: Some(0.2),
+            // XXX: should the default be Never?
```

Comment:
Should the default _of the builder_ be never? If you decide to customize a retry policy, perhaps you expect it to only retry whatever you specified, and a new builder hasn't specified any classifier...

Also, something that might catch you off guard: if you add a classifier, it _changes_ from protocol nacks to whatever you specified, instead of _both_. Or maybe you expect that. But what if you want both? Should we provide an easy option to combine it with a classifier?

Reply:
I've updated it so the default of the _builder_ is "never", and documented as such. Therefore, there's 2 defaults explicitly listed:

- If you don't provide a retry policy, the client uses a default of protocol nacks.
- If you create a retry policy, it doesn't know anything, you need to specify how to classify things.

In support of this, I also added `reqwest::retry::never()` which a user could use to tell a client to stop its default of protocol nacks.